rllib/core/testing/tf/bc_module.py

```diff
@@ -100,3 +101,78 @@ def from_model_config(
         }
 
         return cls(**config)
+
+
+class BCTfRLModuleWithSharedGlobalEncoder(TfRLModule):
+    def __init__(self, encoder, local_dim, hidden_dim, action_dim):
+        super().__init__()
+
+        self.encoder = encoder
+        self.policy_head = tf.keras.Sequential(
+            [
+                tf.keras.layers.Dense(
+                    hidden_dim + local_dim,
+                    input_shape=(hidden_dim + local_dim,),
+                    activation="relu",
+                ),
+                tf.keras.layers.Dense(hidden_dim, activation="relu"),
+                tf.keras.layers.Dense(action_dim),
+            ]
+        )
+
+    def input_specs_inference(self):
+        return self._common_input_spec()
+
+    def input_specs_exploration(self):
+        return self._common_input_spec()
+
+    def input_specs_train(self):
+        return self._common_input_spec()
+
+    def _common_input_spec(self):
+        return [("obs", "global"), ("obs", "local")]
+
+    def _forward_inference(self, batch):
+        return self._common_forward(batch)
+
+    def _forward_exploration(self, batch):
+        return self._common_forward(batch)
+
+    def _forward_train(self, batch):
+        return self._common_forward(batch)
```

Comment:
these too.

Reply:
forward_train() in most cases is not the same as forward_inference or exploration. So I want the user to explicitly implement them. If you look at pytorch_lightning they also have 3 separate functions, even tho they will all need to call forward method of the underlying module. Being explicit here is intentional. For specs I agree. Cuz for 99% of use-cases the main key that should exist is `OBS` at the very least.

Reply:
I see. that makes sense.